Election archiving

NishkaArora opened the pull request:
## Changes Made

### Database Changes
2 new scripts introduced under db/migrate:
db/migrate/20250911073315_add_archived_to_elections.rb
db/migrate/20250911182024_add_election_id_to_visitors.rb

- Added `archived` boolean column to elections table (default: false) - tracks which elections have been archived
- Added `election_id` column to visitors table (default: -1). The migrate script also backfills the election_id column for previous visitors by parsing the url column in the visitor table. The visitor url starts with bases http://localhost:3000/ (local) or with https://pbstanford.org/ (hosted), the string after the base and before the next "/" is compared with the "slug" column in the elections table to find the election_id. 
      - The election_id column is filled for _new_ visitors using the same logic as how the migrate script backfills for previous visitors for consistency (code for tracking election_id for new visitors is in app/controller/application_controller.rb)
- Added indices for both new columns

### Archive button

#### UI additions:
- Red button for "Archive this Election" - only accessible to superadmins
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 53 51 PM" src="https://github.com/user-attachments/assets/4e7c0f0e-bc48-4b3f-9173-da0f444e4796" />

- On clicking the button, you see a confirmation modal. On clicking archive button within the modal, a POST request is sent to the archive endpoint defined in configs/routes.rb, rails routes this to the archive action in ElectionsController.
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 54 04 PM" src="https://github.com/user-attachments/assets/72900b32-bfba-4ae4-b101-207ea50bd67d" />

- Once archived, the "Archive this Election" button changes to "This election has been archived."
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 54 26 PM" src="https://github.com/user-attachments/assets/192c9afe-92e0-42c4-acf7-6a50b2126972" />

- Once archived, the election list has a tag "archived" near the name of the archived elections.
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 54 43 PM" src="https://github.com/user-attachments/assets/33960598-5fd0-461c-8197-79814bc5a6ac" />

- Once archived, voter_registration export is disabled (since the data field has been NULLed)
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 55 17 PM" src="https://github.com/user-attachments/assets/28bed04d-7ac9-4afa-ab94-9eb4e82c099a" />

#### Archive functionality:
- New `archive` action in ElectionsController does the following:
  - Hashes the ip_address and user_agent columns for visitors table and the voters table
  - NULLs voter_registration table's **data** field
  - Anonymizes authentication IDs - ensures anonymized IDs are unique to prevent constraint violations.. The voters table has a unique constraint on the combination of election_id, authentication_method, and authentication_id. This was previously defined in the schema file so that the same access code or phone number cannot be used twice.
  - Sets `archived: true` in the elections table. 
  - Disables "admin config" updates (unchecks the check-box)
       - Validation is added to (app/models/elections.rb) to prevent the "Allow admin to change config and projects" from being enabled again (for archived elections).
 
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 56 08 PM" src="https://github.com/user-attachments/assets/dd86627e-3cd0-4e0a-b8d7-8deb8779e772" />
<img width="1440" height="900" alt="Screenshot 2025-09-12 at 3 56 14 PM" src="https://github.com/user-attachments/assets/2b8cebd6-51d6-493d-86b5-8ebb45f5948e" />

#### Notes
- It is important to stress-test the URL parsing logic for mapping visitor url to election_id. In particular, do we need to add election_id to admins that visit the election page too? Or just the voters that visit the election page? The script will need to be modified for admins that visit pbstanford.org to edit the election.
- todo: The user_agent column in the visitors table contains information about the device used or browser used. Instead of hashing the user_agent column, retaining some information for aggregate stats.

NishkaArora committed:
Election archiving feature. Changes made: Added archive button to election page. Added archived (boolean column) to elections table. Added election_id column to visitors table. Changed script so that election_id is recorded for new visitors. Implemented archive action that anonymizes voter and visitor data. Added validation to prevent admin config changes on archived elections. Updated html to show archived status in the election list.

NishkaArora committed:
small bug fix. Instead of setting the flags in the same function call, we set them separately. The previous way was triggering the validation that prevents changing admin permissions on archived elections

NishkaArora committed:
made a subtle fix so that the checkbox is unchecked by bypassing validation which prevents it from changing later on

ashishgpersonal commented:
Can we please (a) Backup the database before archiving, and (b) If there is any error in url processing, log meaningful information and also let the archiver know that there were errors which are in the log file.

NishkaArora commented:
We checked that the database backs up every 24 hours.

ashishgpersonal commented:
@NishkaArora and @ashishgpersonal verified that there are already daily snapshots being taken by aws. So no need for additional backups.

ashishgpersonal commented on app/controllers/admin/elections_controller.rb:
```diff
@@ -112,6 +112,46 @@ def destroy
+      voters_for_auth_id.find_each do |voter|
+        voter.update_column(:authentication_id, "anonymized_#{voter.id}")
+      end
+      
```
Please create an archived_info table that has election_id as a key and has fields that include date archived, who archived it, and any comments and add a comment field to the archive button so the archiver can write down additional information if they want (but don't make it mandatory to add that information)

StanfordCDT replied:
could we do this in a column of the elections table? Or is there an advantage to do it apart?

lgelauff replied:
could we do this in a column of the elections table? Or is there an advantage to do it apart?

NishkaArora replied:
In the new commits, I changed the "archived" column of the election table from a boolean table to a json table. The default value is null (so all elections are null to start with). When the "archive" method is carried out, we save a json with 3 parameters into the archived column {"username": ABC, "archived_at": date/time, "info": string}

This way we don't need to introduce a new table to the database but we can (if needed) add to the information we collect when archiving. The main downside that comes to mind from not going with the separate-table approach is that we cannot index as easily (such as make SQL queries that select all tables archived by some user). Depending on how important that is and if there are other pros/cons, we can switch to using a separate table.

I have attached an image of the textbox used to input text while archiving.

<img width="1440" height="900" alt="Screenshot 2025-09-30 at 2 02 49 AM" src="https://github.com/user-attachments/assets/907d6bc3-95f8-47ce-98fb-0e7dd660d424" />

lgelauff commented:
Note that this will likely give some errors: 
`Added election_id column to visitors table (default: -1). The migrate script also backfills the election_id column for previous visitors by parsing the url column in the visitor table. The visitor url starts with bases http://localhost:3000/ (local) or with https://pbstanford.org/ (hosted), the string after the base and before the next "/" is compared with the "slug" column in the elections table to find the election_id.`

Elections have had their slug changed, sometimes several times. I dont know if we used to always log these? Given the default, we should be able to resolve iteratively.

lgelauff commented:
"Once archived, the "Archive this Election" button changes to "This election has been archived.""

The screenshot related to this shows 'register in-person voters'. Probably another button we'll want to disable. There are likely some more.

lgelauff commented:
> Once archived, the election list has a tag "archived" near the name of the archived elections.

The screenshot shows the election under 'active elections'. This probably means we forgot something, as an archived election should never be 'active'. This might be that "this election has ended" is not active. It may be a good idea to make this a prerequisite (and consider if there should be other prerequisites/safeguards such as any vote from the past 24h.)

Maybe also good to include a checklist in the popup?
- Did you check if any (test) votes need to be voided? (link to the relevant tab)
- (anything else?)

lgelauff commented:
Please make sure to test: 
- [x] What happens if the process is interrupted mid-way?
- [x] Do the exports still work? (unless disabled)
- [x] Does the preview ballot still work
- [x] Is the codes batch creation disabled? Do we still show those old batches? Should we? 
- [x] Do all analytics still work, including e.g. language stats?

lgelauff commented on db/migrate/20250911182024_add_election_id_to_visitors.rb:
```diff
@@ -0,0 +1,41 @@
+      add_index  :visitors, :election_id
+  
+      # Minimal AR classes scoped to this migration (avoid app models)
+      election_klass = Class.new(ActiveRecord::Base) { self.table_name = "elections" }
```
Is there a reason to use _klass ?

NishkaArora replied:
I believe it is ruby convention for migrate files. I checked that I can change it to class without any keyword conflicts if there is a preference for it.

lgelauff replied:
if it's a convention - no problem. Just curious!

lgelauff commented on db/migrate/20250911182024_add_election_id_to_visitors.rb:
```diff
@@ -0,0 +1,41 @@
+      visitor_klass  = Class.new(ActiveRecord::Base) { self.table_name = "visitors"  }
+  
+      slug_to_id = election_klass.pluck(:slug, :id).to_h
+      bases = ["http://localhost:3000/", "https://pbstanford.org/"]
```
is there a way to avoid hardcoding the domain name?

lgelauff replied:
(sorry, this was stuck in pending! I think this is outdated now)

NishkaArora committed:
election column for archived information, url processing logging, safeguard against archiving active elections, database transaction in case of interruptions, disabled some buttons

NishkaArora committed:
changed the way the election_id is recorded for new visitors to not rely on url and election_slug

NishkaArora commented:
> "Once archived, the "Archive this Election" button changes to "This election has been archived.""
> 
> The screenshot related to this shows 'register in-person voters'. Probably another button we'll want to disable. There are likely some more.

Thank you for catching this! In the new commit, the register in-person voters as well as the convert this computer to a voting machine buttons are disabled after archiving. I also disabled the "generate new codes" and "import codes" buttons. 

I checked and other actions available to admins are already disabled when "allow admins to change config and projects" is disabled (except viewing content/read-only actions). 

When elections are archived, I removed admins' ability to view the previously generated codes. Superadmins may still view them.

<img width="1440" height="900" alt="Screenshot 2025-09-30 at 2 11 21 AM" src="https://github.com/user-attachments/assets/e2a0b23b-afb6-40c5-a7ac-5ec0d6bbdaf3" />
<img width="1440" height="900" alt="Screenshot 2025-09-30 at 2 15 15 AM" src="https://github.com/user-attachments/assets/c520da2a-16f8-4220-b7d2-82d524089bb7" />

NishkaArora commented:
> > Once archived, the election list has a tag "archived" near the name of the archived elections.
> 
> The screenshot shows the election under 'active elections'. This probably means we forgot something, as an archived election should never be 'active'. This might be that "this election has ended" is not active. It may be a good idea to make this a prerequisite (and consider if there should be other prerequisites/safeguards such as any vote from the past 24h.)
> 
> Maybe also good to include a checklist in the popup?
> 
> * Did you check if any (test) votes need to be voided? (link to the relevant tab)
> * (anything else?)

In the latest commit, it is required that an election has ended for a super admin to be able to archive an election:

<img width="1440" height="900" alt="Screenshot 2025-09-30 at 2 10 28 AM" src="https://github.com/user-attachments/assets/5bb522b2-2420-4e63-9bd6-5d6dc9cae267" />
<img width="1440" height="900" alt="Screenshot 2025-09-30 at 2 25 53 AM" src="https://github.com/user-attachments/assets/ab3f1a6c-39ad-4225-9bcb-ba0ac830c161" />

I also added a blurb to the archive modal with a checklist. Let me know if there's anything else I can add to the checklist!

<img width="1440" height="900" alt="Screenshot 2025-09-30 at 2 02 49 AM" src="https://github.com/user-attachments/assets/68407966-c4b8-46af-aa78-f07d7daed889" />

NishkaArora commented:
> Please make sure to test:
> 
> * [x]  What happens if the process is interrupted mid-way?

In the new commits, I added an "ActiveRecord::Base.transaction do - end" around the archiving function which ensures that either all the steps or none of the steps of the archive function are carried out.

> * [x]  Do the exports still work? (unless disabled)

I confirmed that the following exports still work after archiving: export projects, "export to excel" for analytics, export aggregate data and export individual data for approval, export individual data for knapsack, individual/aggregate for ranked and token.

> * [x]  Does the preview ballot still work

I checked that the preview ballot works.

> * [x]  Is the codes batch creation disabled? Do we still show those old batches? Should we?

I disabled the code batch creation and import codes in this commit. It still shows old batches to superadmins and not to admins. Is that a good configuration? I'm not sure how to reason about this.

> * [x]  Do all analytics still work, including e.g. language stats?

- Time and Language stats work
- In the voter table, there is a chart for every vote recorded, that also works, voiding and unvoiding works
- Approval: co-occurence matrix probabilities/counts work, chart and table also work
- Knapsack: checked all the figures
- Token: checked table/chart
- Ranked: table with total scores is displayed

lgelauff commented:
Excellent! It sounds like it functionally does everything I'd hope. The only remaining question is about hardcoding the domain (mostly for the usecase that someone installs it on their own server) - but we can deal with that separately if need be.

ashishgpersonal commented:
Nishka: Can we please create an issue for non hardcoding of the domain so
we don't forget.

Ashish.


On Tue, Sep 30, 2025 at 4:16 AM Lodewijk ***@***.***> wrote:

> *lgelauff* left a comment (StanfordCDT/pb#84)
> <https://github.com/StanfordCDT/pb/pull/84#issuecomment-3351473309>
>
> Excellent! It sounds like it functionally does everything I'd hope. The
> only remaining question is about hardcoding the domain (mostly for the
> usecase that someone installs it on their own server) - but we can deal
> with that separately if need be.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/StanfordCDT/pb/pull/84#issuecomment-3351473309>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGN2ZLGAICT6LXHWRECQNL3VJRBTAVCNFSM6AAAAACGMLNPKCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGNJRGQ3TGMZQHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Ashish Goel

------------------------------------------------
Professor, Departments of Management Science and Engineering, and (by
courtesy) Computer Science,
Stanford University

Cell: 650 814 1478
Office: HEC 359; directions/parking information at
http://www.stanford.edu/~ashishg/directions_work.html .

NishkaArora committed:
moving hard coded URLs to config files

NishkaArora commented:
The only place where we currently use the url-processing is in the migrate file db/migrate/20250911182024_add_election_id_to_visitors.rb when we backfill the election_id column for visitors we've already had in our database. If there are no visitors yet (when someone runs the platform on their own server) then this script will create the empty column election_id in the visitor table. New visitors will have their election_id determined directly without url parsing. 

One way someone running on their own server will run the url parsing is if they manually remove and add back the election_id column from the visitors table (using rails db:rollback and rails db:migrate). In the new commit, I replaced the hard-coded URLs to access the base_urls from the config file of the environment. I tested that it works but it is an extra step for someone running the platform on their own server. Should I include a comment in the README file that when running on their own server, they should update the base_url parameter in the config file?

lgelauff commented:
Right, if it is only backfilling, I'm not very concerned actually. Thanks!

NishkaArora committed:
removing base_urls from config, ruby form helper instead of javascript form, visitor table election_id column default changed to nil, added more info to warning in archive modal

NishkaArora commented:
The latest commit addresses most of Sukolsak's review. The main changes: reverting from base_urls in the config file to hard-coding the urls in the migration file since they are only used in one place, the default value for the election_id column in the visitors table has been changed from -1 to nil, instead of creating the form in the archive modal through javascript we are using a rails helper function, and the modal has a clearer warning about why the archive function is irreversible.

The change that Sukolsak suggested but hasn't been implemented yet is salting/adding a key to the hashing of IP Address.

Here is an image of the new archive modal:
<img width="1440" height="900" alt="Screenshot 2025-10-02 at 10 26 31 PM" src="https://github.com/user-attachments/assets/4c993e19-c22e-4b1d-ad09-4d12d534927d" />

NishkaArora committed:
added a key to hashing in the archive function

NishkaArora commented:
In the latest commit, I added a random key to everything that is hashed in the archive function. The key is not saved anywhere. The same key is used for everything in the election, but a different key would be generated for every election.

NishkaArora committed:
Added time warning to election archive modal

NishkaArora committed:
Added warning banner for admin users for site maintenance

NishkaArora committed:
Added dictionary check to account for changed election slugs when back-filling election id column in visitor table

NishkaArora commented:
DB migrate changes:
- Added a check for election slugs that were changed at some point after collecting visitors to the migrate script that back-fills the visitor table's election_id column. The map from deprecated election slugs to election_id is stored in the migrate script itself.
- Downloaded the complete dataset from the website and tested the archive button on all old elections while checking the logs
- Need advice on this: I ran the migrate script (with backfilling) on the complete dataset locally. There are 6.5 million entries in the visitors table and it took 2.5 hours to finish the backfilling on my computer. The back-filling migrate script runs automatically when we push to Github. However, there is a risk that this will lead to a time-out or lock contention. ChatGPT recommends to separate the creating of the election_id column (in a migrate script) and the backfilling process (in a separate .rake script). This would involve editing the Capfile and I would have to ssh in to run the backfilling. Which one is the better course of action?

UI changes:
- Added an alert to the archive modal explaining that the archiving can take upto 5 minutes depending on the size of the election.
<img width="495" height="738" alt="Screenshot 2025-12-23 at 10 15 45 PM" src="https://github.com/user-attachments/assets/8465bdb3-1f83-435a-90f2-01df10eeb410" />

- Added an easy-to-remove warning banner that the site will be under maintenance.
<img width="1920" height="1080" alt="Screenshot 2025-12-23 at 10 16 45 PM (2)" src="https://github.com/user-attachments/assets/aa770ffb-b3b5-4020-8cd8-5af4bc81feee" />

NishkaArora committed:
Moved the backfilling code from the migration to a rake file that will be called through capistrano to avoid time outs.

NishkaArora committed:
Added a simple progress indicator to the logging in the backfill script.

lgelauff commented:
Suggestion from Claude:
**Suggestion** — `app/controllers/admin/elections_controller.rb`, lines 124–126

The salt is ephemeral and has non-obvious privacy implications worth documenting explicitly.

```ruby
      # Generate a random salt for this archive operation.
      # This salt is used for all SHA-256 hashing (ip_address, user_agent) in this call.
      #
      # IMPORTANT — deliberate design choices:
      #   1. The salt is ephemeral: it is NOT stored anywhere after this request completes.
      #      This means the hashed values are effectively one-way — original IPs/user_agents
      #      cannot be recovered by anyone, including us. This is intentional.
      #   2. The salt is random per-archive: the same IP address hashed in two different
      #      elections will produce two different hashes. Cross-election deduplication or
      #      fraud analysis based on IP/user_agent is permanently impossible after archiving.
      #      This is intentional — we prioritise voter privacy over analytical capability.
      #   3. Do NOT replace this with a stored or global salt. A global salt would allow
      #      re-identification by anyone with access to it, defeating the purpose.
      archive_salt = SecureRandom.hex(32) # 64-character random hex string
```

NishkaArora committed:
Added a warning message for ephemeral archive salt

lgelauff commented:
**Suggestion** — `app/controllers/admin/elections_controller.rb`, archive action: consider bulk SQL for hashing

The current implementation uses `find_each` with individual `update_columns` calls per row. For larger elections this could be slow enough to hit web server request timeouts.

A possible improvement is to replace the row-by-row loops with bulk `UPDATE` statements using MySQL's built-in `SHA2()` function — the database handles the hashing in a single pass:

```ruby
ActiveRecord::Base.transaction do
  # Bulk-hash ip_address and user_agent in one statement per field per table
  [Visitor, Voter].each do |klass|
    klass.where(election_id: election.id)
         .where.not(ip_address: [nil, ''])
         .update_all(["ip_address = SHA2(CONCAT(?, ip_address), 256)", archive_salt])

    klass.where(election_id: election.id)
         .where.not(user_agent: [nil, ''])
         .update_all(["user_agent = SHA2(CONCAT(?, user_agent), 256)", archive_salt])
  end

  VoterRegistrationRecord.where(election_id: election.id).update_all(data: nil)

  # Bulk-anonymize authentication_id
  Voter.where(election_id: election.id)
       .update_all("authentication_id = CONCAT('anonymized_', id)")

  election.update(archived: archived_info)
  election.update_column(:allow_admins_to_update_election, false)
end
```

This reduces the operation to 4–5 SQL statements regardless of election size. The `archive_salt` is passed as a bind parameter so there is no SQL injection risk. Requires MySQL 5.5+ for `SHA2()`, which you already have.

lgelauff reviewed (approved):
Went over this, and you tested it thoroughly. There's some feedback to be implemented after this, but for now, i think it's good enough to go, assuming we lock the site while doing it!

NishkaArora committed:
Added with_lock to election_controller to protect against concurrenct calls to archive through double clicking or multiple admins archiving, improved archived function to check against empty dictionary case.

NishkaArora committed:
Added a maintenance mode that is easy to disable and allows super admins to log in during maintenence.

NishkaArora committed:
edited maintenance mode to render without layout

NishkaArora committed:
Removed the old december maintenance banner.

NishkaArora committed:
changed the text on maintenance banner to 1-2 days instead of few minutes
